src/Auth/Controllers/AccountController.cs

```diff
@@ -0,0 +1,301 @@
+                        return Redirect("~/");
+                    }
+                    else
+                    {
+                        // user might have clicked on a malicious link - should be logged
+                        throw new Exception("invalid return URL");
+                    }
+                }
+
+                await _events.RaiseAsync(new UserLoginFailureEvent(model.Username, "invalid credentials", clientId: context?.ClientId));
+                ModelState.AddModelError(string.Empty, AccountOptions.InvalidCredentialsErrorMessage);
+            }
+
+            // something went wrong, show form with error
+            var vm = await BuildLoginViewModelAsync(model);
+            return View(vm);
+        }
+
+
+        /// <summary>
+        /// Show logout page
+        /// </summary>
+        [HttpGet]
+        public async Task<IActionResult> Logout(string logoutId)
+        {
+            // build a model so the logout page knows what to display
+            var vm = await BuildLogoutViewModelAsync(logoutId);
+
+            if (vm.ShowLogoutPrompt == false)
+            {
+                // if the request for logout was properly authenticated from IdentityServer, then
+                // we don't need to show the prompt and can just log the user out directly.
+                return await Logout(vm);
+            }
+
+            return View(vm);
+        }
+
+        /// <summary>
+        /// Handle logout page postback
+        /// </summary>
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public async Task<IActionResult> Logout(LogoutInputModel model)
+        {
+            // build a model so the logged out page knows what to display
+            var vm = await BuildLoggedOutViewModelAsync(model.LogoutId);
+
+            if (User?.Identity.IsAuthenticated == true)
+            {
+                // delete local authentication cookie
+                await _signInManager.SignOutAsync();
+
+                // raise the logout event
+                await _events.RaiseAsync(new UserLogoutSuccessEvent(User.GetSubjectId(), User.GetDisplayName()));
+            }
+
+            // check if we need to trigger sign-out at an upstream identity provider
+            if (vm.TriggerExternalSignout)
+            {
+                // build a return URL so the upstream provider will redirect back
+                // to us after the user has logged out. this allows us to then
+                // complete our single sign-out processing.
+                string url = Url.Action("Logout", new { logoutId = vm.LogoutId });
+
+                // this triggers a redirect to the external provider for sign-out
+                return SignOut(new AuthenticationProperties { RedirectUri = url }, vm.ExternalAuthenticationScheme);
+            }
+
+            return View("LoggedOut", vm);
+        }
+
+        [HttpGet]
+        public IActionResult AccessDenied()
+        {
+            return View();
+        }
```

Comment:
(Note I accidently deleted this file https://github.com/IdentityServer/IdentityServer4/blob/76dfa41b9648b064b43552190491b917fda40193/samples/Quickstarts/5_EntityFramework/src/IdentityServer/Views/Account/AccessDenied.cshtml so this is a bug on its own)


This one I tried to write something like this:

```
public IActionResult AccessDenied()
        {
            return View("Error", new ErrorViewModel
            {
                Error = new ErrorMessage
                {
                    Error = "Access Denied",
                    ErrorDescription = "You do not have access to this resource.",
                },
            });
        }
```

But then I realized this page doesn't really fit it https://github.com/codidact/core/blob/58e189dd03a7f1d66eb32fc781e696c34d78a6f6/src/Auth/Views/Shared/Error.cshtml

What do you think? Should I bring back the original page or use the generic error?

Reply:
My original suggestion was to tell the Identity framework to go to `/error/403/access-denied` thus leaving the authentication server completely. On a second thought that might not be a good idea after all, mainly because it wouldn't be possible to relay any additional details if there were some.

From the alternatives you proposed, a static error page should be enough. I don't see many other errors that could happen during authentication (a 404 and such aren't handled by the authentication server anyways.)

Reply:
Fixed by https://github.com/codidact/core/pull/29/commits/3e8ca2f42ee1f30799f2ac11339795f2f5e95e77